controllers/controllers/resource/template.go

```diff
@@ -72,16 +76,25 @@ func (r *VsphereTemplate) TemplateResources(ctx context.Context, eksaCluster *an
 		controlPlaneTemplateName = cp.Spec.InfrastructureTemplate.Name
 	}
 
-	var workloadTemplateName string
-	updateWorkloadTemplate := vsphere.AnyImmutableFieldChanged(oldVdc, &vdc, oldWorkerVmc, &workerVmc)
-	if updateWorkloadTemplate {
-		workloadTemplateName = templateBuilder.WorkerMachineTemplateName(clusterName)
-	} else {
-		mcDeployment, err := r.MachineDeployment(ctx, eksaCluster)
-		if err != nil {
-			return nil, err
+	var workloadTemplateNames []string
+	for _, workerNodeGroupConfiguration := range clusterSpec.Spec.WorkerNodeGroupConfigurations {
+		oldVmc := oldWorkerVmcs[workerNodeGroupConfiguration.MachineGroupRef.Name]
+		vmc := workerVmcs[workerNodeGroupConfiguration.MachineGroupRef.Name]
+		updateWorkloadTemplate := vsphere.AnyImmutableFieldChanged(oldVdc, &vdc, &oldVmc, &vmc)
+		if updateWorkloadTemplate {
+			workloadTemplateName := templateBuilder.WorkerMachineTemplateName(clusterName, workerNodeGroupConfiguration.Name)
+			workloadTemplateNames = append(workloadTemplateNames, workloadTemplateName)
+		} else {
+			mcDeployments, err := r.MachineDeployments(ctx, eksaCluster)
+			if err != nil {
+				return nil, err
+			}
+			mdName := fmt.Sprintf("%s-%s", clusterName, workerNodeGroupConfiguration.Name)
+			if _, ok := mcDeployments[mdName]; ok {
+				workloadTemplateName := mcDeployments[mdName].Spec.Template.Spec.InfrastructureRef.Name
+				workloadTemplateNames = append(workloadTemplateNames, workloadTemplateName)
+			}
```

Comment:
handle error case

Reply:
No need to handle the error case. If it doesn't exist, it just won't enter the if block

Reply:
But if it doesn't exist, isn't that a problem? That means there is a template that the worker node group configuration is referencing which isn't actually there?